IDP SAML metadata import fails

**Describe the issue:**
When the IDP SAML metadata contains both the SPSSODescriptor and IDPSSODescriptor,
````
<EntityDescriptor>
  <RoleDescriptor>
  </RoleDescriptor>
  <SPSSODescriptor>
  </SPSSODescriptor>
  <IDPSSODescriptor>
  </IDPSSODescriptor>
</EntityDescriptor>
````
the import fails with the exception below. 
````
ERROR {org.wso2.carbon.idp.mgt.IdentityProviderManager} - Error while updating IdP with name IDP3 org.wso2.carbon.idp.mgt.IdentityProviderManagementException: Error converting metadata
	at org.wso2.carbon.idp.mgt.IdentityProviderManager.handleMetadata(IdentityProviderManager.java:1942)
	at org.wso2.carbon.idp.mgt.IdentityProviderManager.updateIDP(IdentityProviderManager.java:2336)
	at org.wso2.carbon.idp.mgt.IdentityProviderManager.updateIdP(IdentityProviderManager.java:2276)
	at org.wso2.carbon.idp.mgt.IdentityProviderManagementService.updateIdP(IdentityProviderManagementService.java:374)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.axis2.rpc.receivers.RPCUtil.invokeServiceClass(RPCUtil.java:212)
	at org.apache.axis2.rpc.receivers.RPCMessageReceiver.invokeBusinessLogic(RPCMessageReceiver.java:117)
	at org.apache.axis2.receivers.AbstractInOutMessageReceiver.invokeBusinessLogic(AbstractInOutMessageReceiver.java:40)
	at org.apache.axis2.receivers.AbstractMessageReceiver.receive(AbstractMessageReceiver.java:110)
	at org.apache.axis2.engine.AxisEngine.receive(AxisEngine.java:180)
	at org.apache.axis2.transport.local.LocalTransportReceiver.processMessage(LocalTransportReceiver.java:170)
	at org.apache.axis2.transport.local.LocalTransportReceiver.processMessage(LocalTransportReceiver.java:82)
	at org.wso2.carbon.core.transports.local.CarbonLocalTransportSender.finalizeSendWithToAddress(CarbonLocalTransportSender.java:45)
	at org.apache.axis2.transport.local.LocalTransportSender.invoke(LocalTransportSender.java:77)
	at org.apache.axis2.engine.AxisEngine.send(AxisEngine.java:442)
	at org.apache.axis2.description.OutInAxisOperationClient.send(OutInAxisOperation.java:442)
	at org.apache.axis2.description.OutInAxisOperationClient.executeImpl(OutInAxisOperation.java:228)
	at org.apache.axis2.client.OperationClient.execute(OperationClient.java:149)
	at org.wso2.carbon.idp.mgt.stub.IdentityProviderMgtServiceStub.updateIdP(IdentityProviderMgtServiceStub.java:3015)
	at org.wso2.carbon.idp.mgt.ui.client.IdentityProviderMgtServiceClient.updateIdP(IdentityProviderMgtServiceClient.java:315)
	at org.apache.jsp.idpmgt.idp_002dmgt_002dedit_002dfinish_002dajaxprocessor_jsp._jspService(idp_002dmgt_002dedit_002dfinish_002dajaxprocessor_jsp.java:199)
	at org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:71)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:477)
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:385)
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:329)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.wso2.carbon.ui.JspServlet.service(JspServlet.java:207)
	at org.wso2.carbon.ui.TilesJspServlet.service(TilesJspServlet.java:80)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.eclipse.equinox.http.helper.ContextPathServletAdaptor.service(ContextPathServletAdaptor.java:37)
	at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:68)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.wso2.carbon.tomcat.ext.servlet.DelegationServlet.service(DelegationServlet.java:68)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.owasp.csrfguard.CsrfGuardFilter.doFilter(CsrfGuardFilter.java:88)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:126)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.wso2.carbon.tomcat.ext.filter.CharacterSetFilter.doFilter(CharacterSetFilter.java:65)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:126)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.wso2.carbon.identity.context.rewrite.valve.TenantContextRewriteValve.invoke(TenantContextRewriteValve.java:107)
	at org.wso2.carbon.identity.cors.valve.CORSValve.invoke(CORSValve.java:98)
	at org.wso2.carbon.identity.authz.valve.AuthorizationValve.invoke(AuthorizationValve.java:110)
	at org.wso2.carbon.identity.auth.valve.AuthenticationValve.invoke(AuthenticationValve.java:102)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.continueInvocation(CompositeValve.java:99)
	at org.wso2.carbon.tomcat.ext.valves.TomcatValveContainer.invokeValves(TomcatValveContainer.java:49)
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.invoke(CompositeValve.java:62)
	at org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve.invoke(CarbonStuckThreadDetectionValve.java:145)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
	at org.wso2.carbon.tomcat.ext.valves.CarbonContextCreatorValve.invoke(CarbonContextCreatorValve.java:57)
	at org.wso2.carbon.tomcat.ext.valves.RequestEncodingValve.invoke(RequestEncodingValve.java:49)
	at org.wso2.carbon.tomcat.ext.valves.RequestCorrelationIdValve.invoke(RequestCorrelationIdValve.java:126)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:373)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1590)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.wso2.carbon.idp.mgt.IdentityProviderManagementException: Invalid file content
	at org.wso2.carbon.identity.idp.metadata.saml2.util.SAMLMetadataConverter.getFederatedAuthenticatorConfig(SAMLMetadataConverter.java:138)
	at org.wso2.carbon.idp.mgt.IdentityProviderManager.handleMetadata(IdentityProviderManager.java:1907)
	... 83 more
Caused by: org.wso2.carbon.identity.application.common.IdentityApplicationManagementException: No IDP Descriptors found, invalid file content
	at org.wso2.carbon.identity.idp.metadata.saml2.util.SAML2SSOFederatedAuthenticatorConfigBuilder.parse(SAML2SSOFederatedAuthenticatorConfigBuilder.java:73)
	at org.wso2.carbon.identity.idp.metadata.saml2.util.SAML2SSOFederatedAuthenticatorConfigBuilder.build(SAML2SSOFederatedAuthenticatorConfigBuilder.java:343)
	at org.wso2.carbon.identity.idp.metadata.saml2.util.SAMLMetadataConverter.getFederatedAuthenticatorConfig(SAMLMetadataConverter.java:136)
	... 84 more
Caused by: java.lang.ClassCastException: org.opensaml.saml.saml2.metadata.impl.SPSSODescriptorImpl cannot be cast to org.opensaml.saml.saml2.metadata.IDPSSODescriptor
	at org.wso2.carbon.identity.idp.metadata.saml2.util.SAML2SSOFederatedAuthenticatorConfigBuilder.parse(SAML2SSOFederatedAuthenticatorConfigBuilder.java:71)
	... 86 more
````
However, it gets succeeded, if the IDPSSODescriptor takes precedence over SPSSODescriptor.
````
<EntityDescriptor>
  <RoleDescriptor>
  </RoleDescriptor>
  <IDPSSODescriptor>
  </IDPSSODescriptor>
  <SPSSODescriptor>
  </SPSSODescriptor>
</EntityDescriptor>
````

**How to reproduce:**
Import the SAML IDP metadata with the SPSSODescriptor as above.

**Expected behavior:**
The import should be succeeded, extracting the IDPSSODescriptor from the metadata file.

--- a/components/org.wso2.carbon.identity.idp.metadata.saml2/src/main/java/org/wso2/carbon/identity/idp/metadata/saml2/util/SAML2SSOFederatedAuthenticatorConfigBuilder.java
+++ b/components/org.wso2.carbon.identity.idp.metadata.saml2/src/main/java/org/wso2/carbon/identity/idp/metadata/saml2/util/SAML2SSOFederatedAuthenticatorConfigBuilder.java
@@ -62,16 +62,16 @@ public class SAML2SSOFederatedAuthenticatorConfigBuilder {
 
         if (entityDescriptor != null) {
             List<RoleDescriptor> roleDescriptors = entityDescriptor.getRoleDescriptors();
-            //assuming only one IDPSSO is inside the entitydescripter
+            // Assuming that only one IDPSSODescriptor is inside the EntityDescriptor.
             if (CollectionUtils.isNotEmpty(roleDescriptors)) {
-                RoleDescriptor roleDescriptor = roleDescriptors.get(0);
-                if (roleDescriptor != null) {
-                    IDPSSODescriptor idpssoDescriptor;
-                    try {
+                IDPSSODescriptor idpssoDescriptor = null;
+                for (RoleDescriptor roleDescriptor : roleDescriptors) {
+                    if (roleDescriptor instanceof IDPSSODescriptor) {
                         idpssoDescriptor = (IDPSSODescriptor) roleDescriptor;
-                    } catch (ClassCastException ex) {
-                        throw new IdentityApplicationManagementException("No IDP Descriptors found, invalid file content", ex);
+                        break;
                     }
+                }
+                if (idpssoDescriptor != null) {
                     Property properties[] = new Property[24];
 
                     Property property = new Property();
@@ -320,6 +320,8 @@ public class SAML2SSOFederatedAuthenticatorConfigBuilder {
                             }
                         }
                     }
+                } else {
+                    throw new IdentityApplicationManagementException("No IDP Descriptors found, invalid file content.");
                 }
             } else {
                 throw new IdentityApplicationManagementException("No Role Descriptors found, invalid file content");

--- a/components/org.wso2.carbon.identity.sp.metadata.saml2/src/main/java/org/wso2/carbon/identity/sp/metadata/saml2/util/Parser.java
+++ b/components/org.wso2.carbon.identity.sp.metadata.saml2/src/main/java/org/wso2/carbon/identity/sp/metadata/saml2/util/Parser.java
@@ -197,20 +197,21 @@ public class Parser {
         if (entityDescriptor != null) {
             this.setIssuer(entityDescriptor, samlssoServiceProviderDO);
             List<RoleDescriptor> roleDescriptors = entityDescriptor.getRoleDescriptors();
-            //TODO: handle when multiple role descriptors are available
-            //assuming only one SPSSO is inside the entitydescripter
-            RoleDescriptor roleDescriptor;
-            SPSSODescriptor spssoDescriptor;
+            // Assuming that only one SPSSODescriptor is inside the EntityDescriptor.
+            SPSSODescriptor spssoDescriptor = null;
 
             if (CollectionUtils.isEmpty(roleDescriptors)) {
                 throw new InvalidMetadataException("Role descriptor not found.");
             }
-            roleDescriptor = roleDescriptors.get(0);
-
-            if (!(roleDescriptor instanceof SPSSODescriptor)) {
+            for (RoleDescriptor roleDescriptor : roleDescriptors) {
+                if (roleDescriptor instanceof SPSSODescriptor) {
+                    spssoDescriptor = (SPSSODescriptor) roleDescriptor;
+                    break;
+                }
+            }
+            if (spssoDescriptor == null) {
                 throw new InvalidMetadataException("Invalid role descriptor class found.");
             }
-            spssoDescriptor = (SPSSODescriptor) roleDescriptor;
 
             this.setAssertionConsumerUrl(spssoDescriptor, samlssoServiceProviderDO);
             //Response Signing Algorithm - not found

